src/common/oclengine.cpp

```diff
@@ -125,81 +152,122 @@ void OCLEngine::InitOCL()
+        toRet.all_device_contexts[i]->calls[OCL_API_DEC] = cl::Kernel(program, "dec");
+        toRet.all_device_contexts[i]->calls[OCL_API_CDEC] = cl::Kernel(program, "cdec");
+        toRet.all_device_contexts[i]->calls[OCL_API_INCC] = cl::Kernel(program, "incc");
+        toRet.all_device_contexts[i]->calls[OCL_API_DECC] = cl::Kernel(program, "decc");
+        toRet.all_device_contexts[i]->calls[OCL_API_INCS] = cl::Kernel(program, "incs");
+        toRet.all_device_contexts[i]->calls[OCL_API_DECS] = cl::Kernel(program, "decs");
+        toRet.all_device_contexts[i]->calls[OCL_API_INCSC_1] = cl::Kernel(program, "incsc1");
+        toRet.all_device_contexts[i]->calls[OCL_API_DECSC_1] = cl::Kernel(program, "decsc1");
+        toRet.all_device_contexts[i]->calls[OCL_API_INCSC_2] = cl::Kernel(program, "incsc2");
+        toRet.all_device_contexts[i]->calls[OCL_API_DECSC_2] = cl::Kernel(program, "decsc2");
+        toRet.all_device_contexts[i]->calls[OCL_API_INCBCD] = cl::Kernel(program, "incbcd");
+        toRet.all_device_contexts[i]->calls[OCL_API_DECBCD] = cl::Kernel(program, "decbcd");
+        toRet.all_device_contexts[i]->calls[OCL_API_INCBCDC] = cl::Kernel(program, "incbcdc");
+        toRet.all_device_contexts[i]->calls[OCL_API_DECBCDC] = cl::Kernel(program, "decbcdc");
+        toRet.all_device_contexts[i]->calls[OCL_API_INDEXEDLDA] = cl::Kernel(program, "indexedLda");
+        toRet.all_device_contexts[i]->calls[OCL_API_INDEXEDADC] = cl::Kernel(program, "indexedAdc");
+        toRet.all_device_contexts[i]->calls[OCL_API_INDEXEDSBC] = cl::Kernel(program, "indexedSbc");
+        toRet.all_device_contexts[i]->calls[OCL_API_APPROXCOMPARE] = cl::Kernel(program, "approxcompare");
+        toRet.all_device_contexts[i]->calls[OCL_API_NORMALIZE] = cl::Kernel(program, "nrmlze");
+        toRet.all_device_contexts[i]->calls[OCL_API_UPDATENORM] = cl::Kernel(program, "updatenorm");
+        toRet.all_device_contexts[i]->calls[OCL_API_APPLYM] = cl::Kernel(program, "applym");
+        toRet.all_device_contexts[i]->calls[OCL_API_APPLYMREG] = cl::Kernel(program, "applymreg");
+        toRet.all_device_contexts[i]->calls[OCL_API_PHASEFLIP] = cl::Kernel(program, "phaseflip");
+        toRet.all_device_contexts[i]->calls[OCL_API_ZEROPHASEFLIP] = cl::Kernel(program, "zerophaseflip");
+        toRet.all_device_contexts[i]->calls[OCL_API_CPHASEFLIPIFLESS] = cl::Kernel(program, "cphaseflipifless");
+        toRet.all_device_contexts[i]->calls[OCL_API_PHASEFLIPIFLESS] = cl::Kernel(program, "phaseflipifless");
+        toRet.all_device_contexts[i]->calls[OCL_API_MUL] = cl::Kernel(program, "mul");
+        toRet.all_device_contexts[i]->calls[OCL_API_DIV] = cl::Kernel(program, "div");
+        toRet.all_device_contexts[i]->calls[OCL_API_CMUL] = cl::Kernel(program, "cmul");
+        toRet.all_device_contexts[i]->calls[OCL_API_CDIV] = cl::Kernel(program, "cdiv");
+
+        if (saveBinaries) {
```

Comment:
Separate function.

Aim for main() to be able to call “generate” and “save” independently of initializing the system.

Reply:
I'm thinking about how to do this. Basically, though, "build" requires the same process as initializing the environment--we have to identify and initialize contexts and devices, and we have build the programs for them. The only part that could be separated out is the entry point setup for the kernels, and I'm not sure if it's actually useful, to pull just that out.